engine/exchange_manager.go

```diff
@@ -211,3 +247,20 @@ func (m *ExchangeManager) NewExchangeByName(name string) (exchange.IBotExchange,
 	}
 	return exch, nil
 }
+
+// Shutdown shuts down all exchanges and unloads them
+func (m *ExchangeManager) Shutdown() error {
```

Comment:
Depending on context, this function does introduce new issues.
- With websocket shutdown now being included, depending when shutdown is called, causes issues which block the shutdown of the application. Consider the following snippets and their timestamps:
```
[INFO]  | LOG |  17/01/2023 10:00:20  | Captured interrupt, shutdown requested.
```

```
[ERROR] | PORTFOLIO |  17/01/2023 10:00:33  | Error encountered retrieving exchange account info for COINUT. Error COINUT websocket connection: timeout waiting for response with signature: 3785467
[WARN]  | WEBSOCKET |  17/01/2023 10:00:36  | Binance exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:00:52  | Okx exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:01:03  | ZB exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:01:21  | Bybit exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:01:47  | HitBTC exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:02:45  | CoinbasePro exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:02:48  | Huobi exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:02:50  | Bitstamp exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:03:09  | Bithumb exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:03:20  | Bitstamp websocket has been disconnected. Reason: read tcp 192.168.1.13:54344->35.156.126.7:443: use of closed network connection
[WARN]  | EXCHANGE |  17/01/2023 10:03:23  | Bitstamp BTC/USD spot orderbook has zero bid price, filtering.
[ERROR] | WEBSOCKET |  17/01/2023 10:03:23  | Bitstamp cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:03:23  | Bitstamp exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:03:37  | GateIO exchange backlog in websocket processing detected
[ERROR] | WEBSOCKET |  17/01/2023 10:03:40  | Bithumb cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:03:42  | Bithumb exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:03:53  | Bitstamp websocket has been disconnected. Reason: read tcp 192.168.1.13:54365->18.194.85.184:443: use of closed network connection
[WARN]  | EXCHANGE |  17/01/2023 10:03:55  | Bitstamp BTC/USD spot orderbook has zero bid price, filtering.
[ERROR] | WEBSOCKET |  17/01/2023 10:03:55  | Bitstamp cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:03:55  | Bitstamp exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:04:07  | GateIO websocket has been disconnected. Reason: read tcp 192.168.1.13:54340->18.177.40.176:443: use of closed network connection
[ERROR] | WEBSOCKET |  17/01/2023 10:04:09  | GateIO cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:11  | GateIO exchange backlog in websocket processing detected
[ERROR] | WEBSOCKET |  17/01/2023 10:04:14  | Bithumb cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:15  | Bithumb exchange backlog in websocket processing detected
[ERROR] | WEBSOCKET |  17/01/2023 10:04:17  | GateIO subscription failure: GateIO websocket connection: timeout waiting for response with signature: 169499117
[WARN]  | WEBSOCKET |  17/01/2023 10:04:25  | Bitstamp websocket has been disconnected. Reason: read tcp 192.168.1.13:54370->18.194.85.184:443: use of closed network connection
[WARN]  | EXCHANGE |  17/01/2023 10:04:27  | Bitstamp BTC/USD spot orderbook has zero bid price, filtering.
[ERROR] | WEBSOCKET |  17/01/2023 10:04:27  | Bitstamp cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:27  | Bitstamp exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:04:29  | COINUT websocket has been disconnected. Reason: websocket: close 1006 (abnormal closure): unexpected EOF
[ERROR] | WEBSOCKET |  17/01/2023 10:04:31  | COINUT cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:41  | GateIO websocket has been disconnected. Reason: read tcp 192.168.1.13:54373->43.206.6.17:443: use of closed network connection
[ERROR] | WEBSOCKET |  17/01/2023 10:04:42  | GateIO cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:43  | BTC Markets exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:04:43  | GateIO exchange backlog in websocket processing detected
[ERROR] | WEBSOCKET |  17/01/2023 10:04:48  | Bithumb cannot start websocket connection monitor connection monitor is already running
[WARN]  | WEBSOCKET |  17/01/2023 10:04:49  | Bithumb exchange backlog in websocket processing detected
[WARN]  | WEBSOCKET |  17/01/2023 10:04:57  | Bitstamp websocket has been disconnected. Reason: read tcp 192.168.1.13:54375->18.197.135.115:443: use of closed network connection
[ERROR] | WEBSOCKET |  17/01/2023 10:04:57  | GateIO subscription failure: GateIO websocket connection: timeout waiting for response with signature: 149277907, GateIO websocket connection: timeout waiting for response with signature: 189667856
[WARN]  | EXCHANGE |  17/01/2023 10:04:59  | Bitstamp BTC/USD spot orderbook has zero bid price, filtering.
```

I had to forcefully quit the application after 10 minutes

Reply:
- If you do time the application shutdown correctly, you get some extra verbosity that I think we could probably do without:
```
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | ZB websocket has been disconnected. Reason: read tcp 192.168.1.13:54249->172.67.31.157:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | ZB exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | GateIO websocket has been disconnected. Reason: read tcp 192.168.1.13:54259->18.177.40.176:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | GateIO exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Okx websocket has been disconnected. Reason: read tcp 192.168.1.13:54246->104.18.2.151:8443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Okx exchange unloaded successfully.
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Lbank exchange unloaded successfully.
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | BTSE exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | BTC Markets websocket has been disconnected. Reason: read tcp 192.168.1.13:54248->172.64.150.138:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | BTC Markets exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Gemini websocket has been disconnected. Reason: read tcp 192.168.1.13:54255->18.211.85.240:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Gemini exchange unloaded successfully.
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | LocalBitcoins exchange unloaded successfully.
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Yobit exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Poloniex websocket has been disconnected. Reason: read tcp 192.168.1.13:54253->104.16.116.135:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Poloniex exchange unloaded successfully.
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | ITBIT exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Binanceus websocket has been disconnected. Reason: read tcp 192.168.1.13:54267->34.238.66.24:9443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Binanceus exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Huobi websocket has been disconnected. Reason: read tcp 192.168.1.13:54250->104.16.230.188:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Huobi exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Bitstamp websocket has been disconnected. Reason: read tcp 192.168.1.13:54265->18.197.135.115:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Bitstamp exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Binance websocket has been disconnected. Reason: read tcp 192.168.1.13:54260->54.199.160.150:9443: use of closed network connection
[ERROR] | LOG |  17/01/2023 09:59:01  | HitBTC failed to get cryptocurrency deposit address for DASH. Err: HitBTC unsuccessful HTTP status code: 403 raw response: {"error":{"code":1003,"message":"Action is forbidden for this API key"}}
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Binance exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | CoinbasePro websocket has been disconnected. Reason: read tcp 192.168.1.13:54245->104.18.8.210:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | CoinbasePro exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Bitfinex websocket has been disconnected. Reason: read tcp 192.168.1.13:54261->104.16.173.181:443: use of closed network connection
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Bitfinex websocket has been disconnected. Reason: read tcp 192.168.1.13:54270->104.16.175.181:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Bitfinex exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | OKCOIN International websocket has been disconnected. Reason: read tcp 192.168.1.13:54264->104.18.158.225:8443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | OKCOIN International exchange unloaded successfully.
[WARN]  | WEBSOCKET |  17/01/2023 09:59:01  | Bitmex websocket has been disconnected. Reason: read tcp 192.168.1.13:54254->172.64.146.66:443: use of closed network connection
[INFO]  | EXCHANGE |  17/01/2023 09:59:01  | Bitmex exchange unloaded successfully.
[ERROR] | WEBSOCKET |  17/01/2023 09:59:01  | Bithumb cannot start websocket connection monitor connection monitor is already running
[INFO]  | LOG |  17/01/2023 09:59:01  | Exiting
```

Reply:
- I would consider the introduction of asynchronous processing due the time it can take to shutdown a websocket

Reply:
The reason I brought up context was that when shutting down the application, we really don't actually need to wait for the websocket to shut down correctly. Only when shutting down our own exchange manager, so you may wish to introduce a new function, or introduce a _force_ flag or something along those lines